chore: upgrade to jest 24

SimenB opened the pull request:
This won't work on CI since `pretty-format` uses `react-is` internally, and since yarn finds `react-is` in this repo, it won't install it. I don't know how to tell yarn to install a version of `react-is` into `pretty-format`s `node_modules`? To test this locally, I just did `cp -R ../jest/node_modules/react-is node_modules/pretty-format/node_modules` which worked fine. Maybe @arcanis can help?

All tests pass locally 🙂 

---

~This contains #15779 with the upgrade to jest 24 applied on top of it~

SimenB commented on package.json:
```diff
@@ -4,12 +4,12 @@
     "babel-code-frame": "^6.26.0",
     "babel-core": "^6.0.0",
     "babel-eslint": "^10.0.0",
-    "babel-jest": "^23.0.1",
```
you don't use this, you have your own transformer

SimenB commented on packages/eslint-plugin-react-hooks/__tests__/ESLintRuleExhaustiveDeps-test.js:
```diff
@@ -521,7 +521,7 @@ const tests = {
       code: `
         function useMyThing(myRef) {
-          const myRef = useRef();
+          myRef = useRef();
```
syntax error raised by babel since `myRef` existed already

SimenB commented on packages/eslint-plugin-react-hooks/__tests__/ESLintRulesOfHooks-test.js:
```diff
@@ -95,7 +95,7 @@ eslintTester.run('react-hooks', ReactHooksESLintRule, {
       ({useHook: () => { useState(); }});
       ({useHook() { useState(); }});
-      const {useHook = () => { useState(); }} = {};
+      const {useHook3 = () => { useState(); }} = {};
```
syntax error raised by babel since `useHook` existed already

SimenB commented on packages/react-dom/src/__tests__/ReactDOMServerIntegrationElements-test.js:
```diff
@@ -648,7 +648,7 @@ describe('ReactDOMServerIntegration', () => {
     });
 
-    describe('component hierarchies', async function() {
+    describe('component hierarchies', function() {
```
`describe` must return `void`

SimenB commented on packages/react-test-renderer/src/__tests__/ReactTestRendererAsync-test.js:
```diff
@@ -160,7 +160,7 @@ describe('ReactTestRendererAsync', () => {
       expect(() =>
         expect(Scheduler).toFlushAndYieldThrough(['foo', 'baz']),
-      ).toThrow('Expected value to equal:');
+      ).toThrow('// deep equality');
```
we've changed the error messages

SimenB commented on packages/react/src/__tests__/ReactCoffeeScriptClass-test.coffee:
```diff
@@ -543,3 +543,4 @@ describe 'ReactCoffeeScriptClass', ->
     node = ReactDOM.findDOMNode(instance)
     expect(node).toBe container.firstChild
     undefined
+  undefined
```
same as above, `describe` must return `undefined`

SimenB commented on packages/react/src/__tests__/createReactClassIntegration-test.js:
```diff
@@ -453,7 +453,7 @@ describe('create-react-class-integration', () => {
       ReactDOM.render(<Foo foo="foo" />, document.createElement('div')),
     ).toWarnDev(
-      'Component: getDerivedStateFromProps() is defined as an instance method ' +
+      'Foo: getDerivedStateFromProps() is defined as an instance method ' +
```
probably due to us having source maps now

SimenB replied:
nope, just babel 7 being better 🙂

SimenB commented on scripts/jest/config.base.js:
```diff
@@ -12,12 +12,13 @@ module.exports = {
   roots: ['<rootDir>/packages', '<rootDir>/scripts'],
   collectCoverageFrom: ['packages/**/*.js'],
   timers: 'fake',
+  testURL: 'about:blank',
```
the default changed in some Jest release from `about:blank` to `http://localhost` - that broke a test you had for `href`

SimenB commented on scripts/jest/config.base.js:
```diff
@@ -12,12 +12,13 @@ module.exports = {
   },
   setupFiles: [require.resolve('./setupEnvironment.js')],
-  setupTestFrameworkScriptFile: require.resolve('./setupTests.js'),
+  setupFilesAfterEnv: [require.resolve('./setupTests.js')],
```
renamed option

SimenB commented on scripts/jest/preprocessor.js:
```diff
@@ -53,6 +53,9 @@ module.exports = {
     if (filePath.match(/\.ts$/) && !filePath.match(/\.d\.ts$/)) {
       return tsPreprocessor.compile(src, filePath);
     }
+    if (filePath.match(/\.json$/) ) {
```
Jest passes json files through transformers now, and babel didn't like it 🙂

SimenB commented on scripts/jest/preprocessor.js:
```diff
@@ -70,7 +73,7 @@ module.exports = {
               }
             : {}
         )
-      ).code;
```
this gives jest source maps from babel

arcanis commented:
> since yarn finds `react-is` in this repo, it won't install it. I don't know how to tell yarn to install a version of `react-is` into `pretty-format`s `node_modules`?

Not sure I understand - you mean that `/node_modules/react-is` exists, so Yarn doesn't create `/node_modules/pretty-format/node_modules/react-is`? Yarn should only do this if the version at the top-level is compatible with the version requested by `pretty-format`. Isn't it the case?

SimenB commented:
>Not sure I understand - you mean that `/node_modules/react-is` exists, so Yarn doesn't create `/node_modules/pretty-format/node_modules/react-is`?

Correct.

>Yarn should only do this if the version at the top-level is compatible with the version requested by `pretty-format`. Isn't it the case?

It is semver compatible, but the package has not been built/compiled to cjs so it doesn't work. Its main point to this file: https://github.com/facebook/react/blob/3b2302253f13c9cd049fac10784e10b6c582e3b6/packages/react-is/index.js

I'd like to tell yarn to ignore the `react-is` from the workspace and install it into `pretty-format` (at least, I think that's what I want 😀)

arcanis commented:
Oh I see, so you want to ignore the local workspace .. hm .. this should be doable with [resolutions](https://yarnpkg.com/en/docs/selective-version-resolutions). Something like this, maybe?

```json
{
  "resolutions": {
    "pretty-format/react-is": "npm:react-is@<insert the semver range>"
  }
}
```

SimenB commented:
That sorta works - it does indeed install from the npm registry, but it installs it to `node_modules/react-is` meaning the one in the workspace is not there. That might break other code importing it (via some haste resolution or some such)?

arcanis commented:
Damn, that's likely a bug in the `resolutions` implementation (fwiw we have tests on this in the v2 trunk, so I know it works properly there) 😢

I don't see an easy fix then ... maybe a `resolutions` entry on `pretty-format` instead, to point to a local copy that would list `react-is@npm:^<range>`? It starts to be a bit hacky 🤔

It's interesting we never got this problem before, though!

threepointone commented:
Now's a good time for us to land this. What do you folks think is the best option right to resolve this pretty-format/react-is issue?

SimenB commented:
I'm not sure :( I didn't try out @arcanis's last suggestion since he said it was hacky. Maybe give that a go? I'm on vacation for a few more days, so if you have time, that'd be great. Otherwise I'll give it a whack some time next week

(Landing #15779 in the meantime would be great, to reduce this diff)

threepointone commented:
Cool cool, enjoy your vacation! I’ll poke at this myself soon

SimenB commented:
I've rebased this now that the other PR landed, but it still has the same `react-is` issue. Not sure how to deal with it, tbh

bvaughn commented:
Figuring out a way to land this might make #266 a little easier to land, but I'm going to try to forge forward and not block on this. 😄

SimenB commented:
I played around some more with this, and got nowhere 😅 One thing that _works_ is https://github.com/facebook/react/pull/15778#issuecomment-497321320, but as mentioned that means that any other module in this monorepo that does `require('react-is');` will get the one from npm instead of the local one. Grepping shows this only happens in one spot, but still: https://github.com/facebook/react/blob/6f86294e686e172fb8da32880f70fa038d644c74/packages/react-test-renderer/src/ReactShallowRenderer.js#L11.

I'll push a commit with it. Thoughts on `moduleNameMapper` in jest config to point `react-is` to a local one? Then at least for tests you'll get the local version

SimenB commented:
Bah, that broke the build... Tests passed, though! I don't know how to solve this beyond manually installing `react-is` from npm into `node_modules/pretty-format/node_modules` via `postinstall`. Or use yarn 2 if it's ready? It's in RC as of 3 hours ago, that means "stable" in JS land, right?

https://github.com/yarnpkg/berry/releases/tag/2019-08-16

arcanis commented:
Fwiw I don't think the React repo will be able to switch unless Flow adds for in-zip loading, or someone writes a linker plugin for Yarn 2 that basically unplugs everything (which is doable).

Feel free to ping me on the Yarnpkg discord channel if you want to discuss it further.

SimenB commented:
That last commit is about as far from yarn v2 compatible you could get I think :P CI passed, though! :tada:

I think I've spent as much time as I'm willing to on this now, unless someone comes up with any fresh new ideas

SimenB commented on scripts/yarn/downloadReactIsForPrettyFormat.js:
```diff
@@ -0,0 +1,30 @@
+
+rimraf.sync(prettyFormatNodeModulesReactIsDir);
+
+pacote
```
could possibly check in the tarball to avoid the network hit? Not sure if you want this at all, so I went with what's easiest/quickest/shortest

bvaughn commented:
> One thing that works is #15778 (comment), but as mentioned that means that any other module in this monorepo that does require('react-is'); will get the one from npm instead of the local one. Grepping shows this only happens in one spot

The new DevTools package (#16381) uses it too, unfortunately.

SimenB commented:
My current solution should work for it - I didn't go with the linked one 🙂

SimenB commented:
Or did you mean that devtools requires `react-is` normally, and fails somehow? If so the same approach as in this PR should work

SimenB commented:
@threepointone thoughts on this?

SimenB commented:
rebased

threepointone commented:
I’ll bring this up in our weekly sync this week, we’ll take a resolution on it one way or the other.

SimenB commented:
Sounds good! Looking forward to it (hopefully) landing 😀

arcanis commented:
Btw, wouldn't it be simpler to just write this one module in standard js? It sounds quite small.

Another thought - would it work to run Jest through babel-register, configuring it to only transpile this one folder?

SimenB commented:
Can't really comment on the first one. As for babel register though, I _think_ it wouldn't work when jest spins up workers to run tests.

---

I have no idea how I broke danger...

SimenB commented:
OK, I broke danger by giving it babel 7 instead of 6. One quickfix is

```diff
diff --git c/package.json w/package.json
index db625cd6e..90561d8d5 100644
--- c/package.json
+++ w/package.json
@@ -34,6 +34,7 @@
     "@babel/traverse": "^7.0.0",
     "@mattiasbuelens/web-streams-polyfill": "0.1.0",
     "art": "^0.10.1",
+    "babel-core": "^6.26.3",
     "babel-eslint": "^10.0.0",
     "babel-plugin-syntax-trailing-function-commas": "^6.5.0",
     "babylon": "6.18.0",
diff --git c/yarn.lock w/yarn.lock
index 86804bbbf..83f231c6b 100644
--- c/yarn.lock
+++ w/yarn.lock
@@ -2100,7 +2100,7 @@ babel-code-frame@^6.16.0, babel-code-frame@^6.26.0:
     esutils "^2.0.2"
     js-tokens "^3.0.2"
 
-babel-core@^6.26.0, babel-core@^6.7.2:
+babel-core@^6.26.0, babel-core@^6.26.3, babel-core@^6.7.2:
   version "6.26.3"
   resolved "https://registry.yarnpkg.com/babel-core/-/babel-core-6.26.3.tgz#b2e2f09e342d0f0c88e2f02e067794125e75c207"
   dependencies:
```

however, the real fix is probably to upgrade danger - v3.0.4 in this repo is 6 majors behind

SimenB commented:
🎉

SimenB commented:
Rebased and squashed the extra commits to migrate devtools over. So now it's one commit migrating to v24, and the one commit I'm rather hesitant to have my name next to as a separate one for the hacky `react-is` workaround. Should make it easier to review at least!

(I often use this repo to test out breaking changes in Jest itself, especially around fake timers, which is why I wanna see it upgraded. 🙂 Makes it easier)

SimenB commented on packages/react-devtools-extensions/package.json:
```diff
@@ -25,7 +25,6 @@
     "archiver": "^3.0.0",
     "babel-core": "^7.0.0-bridge",
     "babel-eslint": "^9.0.0",
-    "babel-jest": "^24.7.1",
```
these are just remnants after migrating it into this monorepo, I guess? They're unused

SimenB commented on packages/react-devtools-extensions/package.json:
```diff
@@ -25,7 +25,6 @@
     "archiver": "^3.0.0",
     "babel-core": "^7.0.0-bridge",
```
this is quite possible also redundant now? not sure... but if it was here for jest@23, then it can go 🙂

SimenB commented:
@threepointone any news from your meeting?

SimenB commented:
@threepointone ping 🙂

threepointone commented:
I've approved, but could you rebase one last time and make sure nothing's broken? I don't expect it to be, but want to be certain before I hit merge :)

SimenB commented:
Awesome! I'll give it a rebase tonight or tomorrow

SimenB commented:
rebased, and CI still seems happy 👍

threepointone commented:
bing bang bosh, it's in!

SimenB commented:
woo, thanks! 🎉

threepointone commented:
No, thank you! We can finally start using inline snapshots yay 😃
